File: Partition List.py
# Definition for singly-linked list.
# class ListNode:
#	 def __init__(self, x):
#		 self.val = x
#		 self.next = None

class Solution:
	# @param head, a ListNode
	# @param x, an integer
	# @return a ListNode
	def partition(self, head, x):
		lHead = None
		lTail = None
		geHead = None
		geTail = None
		node = head
		while (node):
			if (node.val < x):
				if (not lHead):
					lHead = lTail = node
				else:
					lTail.next = node
					lTail = node
			else:
				if (not geHead):
					geHead = geTail = node
				else:
					geTail.next = node
					geTail = node
			temp = node
			node = node.next
			temp.next = None
		if (not lHead):
			return geHead
		lTail.next = geHead
		return lHead

File: Partition List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode lHead = null;
        ListNode lTail = null;
        ListNode geHead = null;
        ListNode geTail = null;
        ListNode node = head;
        while (node != null) {
            if (node.val < x) {
                if (lHead == null) {
                    lHead = lTail = node;
                }
                else {
                    lTail.next = node;
                    lTail = node;
                }
            }
            else {
                if (geHead == null) {
                    geHead = geTail = node;
                }
                else {
                    geTail.next = node;
                    geTail = node;
                }
            }
            ListNode cutTail = node;
            node = node.next;
            cutTail.next = null;
        }
        if (lHead == null)
            return geHead;
        lTail.next = geHead;
        return lHead;
    }
}